### PR TITLE
gs2200m: Check disassociation from AP and enable rejoin

### PR DESCRIPTION
**Pull request by TE-Masatoshi-Ueno**
Summary:
- This commit enables gs2200m driver re-associate to the access point when disassociated

Impact:
- Affects gs2200m only

Testing:
- Tested with spresense:wifi

@masayuki2009

### CODE REVIEW

File: drivers/wireless/gs2200m.c [hunk 1]
```diff
@@ -3071,6 +3086,34 @@ static void gs2200m_irq_worker(FAR void *arg)
...
+      /* Disassociate recovery */
+
+      wlwarn("=== receive DISASSOCIATE\n");
+      dev->valid_cid_bits = 0;
+
+      gs2200m_disassociate(dev);
+      do
+        {
+          while (gs2200m_recv_pkt(dev, pkt_dat) != TYPE_TIMEOUT)
+            {
```
Comment (masayuki2009) [inline drivers/wireless/gs2200m.c:3100]: Do we have more packets to receive?

**Review by masayuki2009 — CHANGES_REQUESTED** [review]

**Comment by masayuki2009** [comment]
@TE-Masatoshi-Ueno 
How about existing connections and packet queue?
In my understanding, the connections will be invalid after rejoining to AP.

File: drivers/wireless/gs2200m.c [hunk 2]
```diff
@@ -247,12 +248,17 @@ static const struct file_operations g_gs2200m_fops =
 static struct evt_code_s _evt_table[] =
 {
   {"OK", TYPE_OK},
+  {"Disassociation Event", TYPE_DISASSOCIATE},
+  {"DISASSOCIATED", TYPE_DISASSOCIATE},
```
[the remark this answers is not in the document]
Comment (TE-Masatoshi-Ueno) [reply drivers/wireless/gs2200m.c:252]: In most cases we will receive "Disassociation Event" message. We will receive "DISASSOCIATED" message only when we don't associate to AP and send connect AT command, so I should remove "DISASSOCIATED" message from this table.

File: drivers/wireless/gs2200m.c [hunk 3]
```diff
@@ -247,12 +248,17 @@ static const struct file_operations g_gs2200m_fops =
...
   {"ERROR", TYPE_ERROR},
   {"DISCONNECT", TYPE_DISCONNECT},
   {"CONNECT", TYPE_CONNECT},
   {"Serial2WiFi APP", TYPE_BOOT_MSG}
 };
 
+static bool gs2200m_disassociate_flag = false;
+static struct gs2200m_assoc_msg reconnect_msg;
+
```
[the remark this answers is not in the document]
Comment (TE-Masatoshi-Ueno) [reply drivers/wireless/gs2200m.c:261]: OK, I will move to struct gs2200m_dev_s.

File: drivers/wireless/gs2200m.c [hunk 4]
```diff
@@ -3071,6 +3086,34 @@ static void gs2200m_irq_worker(FAR void *arg)
...
 
+  if (true == gs2200m_disassociate_flag)
+    {
+      /* Disassociate recovery */
+
+      wlwarn("=== receive DISASSOCIATE\n");
+      dev->valid_cid_bits = 0;
+
+      gs2200m_disassociate(dev);
+      do
```
[the remark this answers is not in the document]
Comment (TE-Masatoshi-Ueno) [reply drivers/wireless/gs2200m.c:3097]: I add this command for insurance, but find this function send disassociation AT command only...  Yes, I should remove this.

File: drivers/wireless/gs2200m.c (hunk 1 above)
Comment (TE-Masatoshi-Ueno) [reply drivers/wireless/gs2200m.c:3100]: If some socket programs send packets before disassociated, many "ERROR: INVALID INPUT" response is received in this timing.

**Comment by TE-Masatoshi-Ueno** [comment]
Yes, all connections are reset after disassociated. so, all cid reset by this command, "dev->valid_cid_bits = 0;".

File: drivers/wireless/gs2200m.c (hunk 1 above)
Comment (masayuki2009) [reply drivers/wireless/gs2200m.c:3100]: Hmm, so you mean that we might receive "ERROR: INVALID INPUT"  messages AFTER we received "Disassociation Event".  So here, we need to call gs2200m_recv_pkt() to discard such messages UNTIL timeout happens.
Is my understanding correct?

Comment (TE-Masatoshi-Ueno) [reply drivers/wireless/gs2200m.c:3100]: Yes,  that's right.

File: drivers/wireless/gs2200m.c [hunk 5]
```diff
@@ -2573,6 +2584,8 @@ static int gs2200m_ioctl_assoc_sta(FAR struct gs2200m_dev_s *dev,
 {
   enum pkt_type_e t;
 
+  memcpy(&reconnect_msg, msg, sizeof(struct gs2200m_assoc_msg));
+
```
[the remark this answers is not in the document]
Comment (TE-Masatoshi-Ueno) [reply drivers/wireless/gs2200m.c:2588]: Please confirm commit 04db01d.

File: drivers/wireless/gs2200m.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (TE-Masatoshi-Ueno) [reply drivers/wireless/gs2200m.c:3096]: Please confirm commit 7577186 .

File: drivers/wireless/gs2200m.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (TE-Masatoshi-Ueno) [reply drivers/wireless/gs2200m.c]: Please confirm commit a4d35d8 .

File: drivers/wireless/gs2200m.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (TE-Masatoshi-Ueno) [reply drivers/wireless/gs2200m.c]: Please confirm commit a4d35d8 .

File: drivers/wireless/gs2200m.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (masayuki2009) [reply drivers/wireless/gs2200m.c:2588]: OK, but please fixup the commit (we don't need a separate commit)

File: drivers/wireless/gs2200m.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (masayuki2009) [reply drivers/wireless/gs2200m.c]: OK, but please fixup the commit (we don't need the separate commit)

File: drivers/wireless/gs2200m.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (masayuki2009) [reply drivers/wireless/gs2200m.c]: OK, but please fixup the commit (we don't need the separate commit)

File: drivers/wireless/gs2200m.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (masayuki2009) [reply drivers/wireless/gs2200m.c:3096]: OK, but please fixup the commit (we don't need the separate commit)

File: drivers/wireless/gs2200m.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (TE-Masatoshi-Ueno) [reply drivers/wireless/gs2200m.c]: Thanks. I merged into 1 commit 928c5ce .

File: drivers/wireless/gs2200m.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (TE-Masatoshi-Ueno) [reply drivers/wireless/gs2200m.c]: Thanks. I merged into 1 commit 928c5ce .

File: drivers/wireless/gs2200m.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (TE-Masatoshi-Ueno) [reply drivers/wireless/gs2200m.c:2588]: Thanks. I merged into 1 commit 928c5ce .

File: drivers/wireless/gs2200m.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (TE-Masatoshi-Ueno) [reply drivers/wireless/gs2200m.c:3096]: Thanks. I merged into 1 commit 928c5ce .

File: drivers/wireless/gs2200m.c (hunk 1 above)
Comment (TE-Masatoshi-Ueno) [reply drivers/wireless/gs2200m.c:3099]: I checked this situation. When gs2200m is disassociated, gs2200m resets all connections. But driver don't catch this condition immediately and keep sending and retrying bulk packet for a while. So. gs2200m respond "ERROR: INVALID INPUT" text message in this timing.

File: drivers/wireless/gs2200m.c [hunk 6]
```diff
@@ -3071,6 +3086,33 @@ static void gs2200m_irq_worker(FAR void *arg)
...
+        {
+          while (gs2200m_recv_pkt(dev, pkt_dat) != TYPE_TIMEOUT)
+            {
+              /* release & initialize pkt_dat before retry */
+
+              _release_pkt_dat(dev, pkt_dat);
+              memset(pkt_dat, 0, sizeof(pkt_dat));
+
+              usleep(100 * 1000);
+            }
```
[the remark this answers is not in the document]
Comment (TE-Masatoshi-Ueno) [reply drivers/wireless/gs2200m.c:3103]: I update to gs2200m_recv_pkt(dev, NULL) and nxsig_usleep(), and add comment at 76f2069 .

[the remark this answers is not in the document]
Comment (masayuki2009) [reply drivers/wireless/gs2200m.c:3103]: Thanks.

File: drivers/wireless/gs2200m.c (hunk 1 above)
Comment (masayuki2009) [reply drivers/wireless/gs2200m.c]: OK, I understand.

File: drivers/wireless/gs2200m.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (masayuki2009) [reply drivers/wireless/gs2200m.c:3103]: Thanks.

**Review by masayuki2009 — APPROVED** [review]
LGTM

**Pull request merged**